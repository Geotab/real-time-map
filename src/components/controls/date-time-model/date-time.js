import storage from "../../../dataStore";
import { getLiveTime } from "../../../utils/helper";
import { showSnackBar } from "../../snackbar/snackbar";

export const dateTimeModel = {

   get dateInput() {
      return document.getElementById("dateInputBox");
   },

   get startTimeInput() {
      return document.getElementById("timeRangeStart");
   },

   get currentTimeInput() {
      return document.getElementById("currentTimeInput");
   },

   initDateTimeInput() {

      this.setDefaultDateValue();
      this.setDefaultStartTimeValue();
      this.setDefaultCurrentTimeValue();

      this.dateInput.addEventListener("keyup", this.onDateEntered.bind(this));
      this.startTimeInput.addEventListener("keyup", this.onStartTimeEntered.bind(this));
      this.currentTimeInput.addEventListener("keyup", this.onCurrentTimeEntered.bind(this));

      this.dateInput.addEventListener("change", onChange);
      this.startTimeInput.addEventListener("change", onChange);
      this.currentTimeInput.addEventListener("change", onChange);

      storage.dateKeeper$.subscribe(this.updateCurrentSecond.bind(this));
   },

   setDefaultDateValue() {
      this.dateInput.value = new Date(storage.currentTime).toLocaleDateString("en-CA");
   },

   setDefaultStartTimeValue() {
      this.startTimeInput.value = new Date(storage.timeRangeStart).toLocaleTimeString("en-GB");
   },

   setDefaultCurrentTimeValue() {
      this.currentTimeInput.value = new Date(storage.currentTime).toLocaleTimeString("en-GB");
   },

   updateCurrentSecond(currentSecond) {

      if (this.dateInput && document.activeElement != this.dateInput) {
         this.setDefaultDateValue();
      }

      if (this.startTimeInput && document.activeElement != this.startTimeInput) {
         this.setDefaultStartTimeValue();
      }

      if (this.currentTimeInput && document.activeElement != this.currentTimeInput) {
         this.currentTimeInput.value = new Date(currentSecond).toLocaleTimeString("en-GB");
      }
   },


   onDateEntered(event) {
      if (!checkRightKeyEntered(event)) {
         return;
      }

      // this.dateInput.blur();
      const selectedTime = new Date(this.dateInput.value + " " + this.currentTimeInput.value);

      if (checkDateInFuture(selectedTime)) {
         this.setDefaultDateValue();
      } else {
         this.dateInput.value = selectedTime.toLocaleDateString("en-CA");
         this.applyAndUpdate(selectedTime.getTime());
      }
   },

   onStartTimeEntered(event) {
      if (!checkRightKeyEntered(event)) {
         return;
      }

      this.startTimeInput.blur();
      const selectedTime = new Date(this.dateInput.value + " " + this.startTimeInput.value);

      if (checkDateInFuture(selectedTime)) {
         this.setDefaultStartTimeValue();

      } else {
         const newTime = selectedTime.getTime();
         storage.timeRangeStart = newTime;
         if (newTime > storage.currentTime) {
            storage.currentTime = newTime;
         }
         this.applyAndUpdate(storage.currentTime);
      }
   },

   onCurrentTimeEntered(event) {
      if (!checkRightKeyEntered(event)) {
         return;
      }

      this.currentTimeInput.blur();
      const selectedTime = new Date(this.dateInput.value + " " + this.currentTimeInput.value);

      if (checkDateInFuture(selectedTime)) {
         this.setDefaultCurrentTimeValue();

      } else {
         const newTime = selectedTime.getTime();
         if (newTime < storage.timeRangeStart) {
            storage.timeRangeStart = newTime;
         }
         this.applyAndUpdate(newTime);
      }
   },

   applyAndUpdate(newTime) {
      this.updateCurrentSecond(newTime);
      storage.dateKeeper$.setNewTime(newTime);
   }
};

function checkRightKeyEntered(event) {
   event.preventDefault();
   // Number 13 is the "Enter" key on the keyboard
   return (event.keyCode === 13);
}

function checkDateInFuture(selectedTime) {

   const liveDate = new Date(getLiveTime());

   if (selectedTime > liveDate) {
      showSnackBar("Selected date is in future!");
      return true;
   }

   return false;
}

export function onChange() {
   showSnackBar("Please hit Enter to apply your changes!");
};