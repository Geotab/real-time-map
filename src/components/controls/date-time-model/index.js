export { dateTimeModel } from "./date-time";