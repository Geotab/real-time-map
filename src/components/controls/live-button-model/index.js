export { liveButtonModel } from "./live-button-model";