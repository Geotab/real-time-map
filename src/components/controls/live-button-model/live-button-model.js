import storage from "../../../dataStore/index";
import { checkIfLive, getLiveTime } from "../../../utils/helper";

export const liveButtonModel = {
  liveButton: undefined,
  liveText: undefined,
  islive: true,

  initLiveButton() {
    this.liveButton = document.getElementById("RTM-LiveDot");
    this.liveText = document.getElementById("RTM-LiveButton");
    this.setLiveBackground();
    storage.dateKeeper$.subscribe(this.checkIfLive.bind(this));
  },

  checkIfLive(currentSecond) {
    
    const timeAhead = checkIfLive(currentSecond);

    if (timeAhead) {
      if (timeAhead > 1800) {
        this.islive = false;
        this.goToLive();
        return;
      }

      if (!this.islive) {
        storage.dateKeeper$.setPeriod(1000);
        this.islive = true;
        this.setLiveBackground();
      }
      return true;

    } else {

      if (this.islive) {
        this.islive = false;
        this.setNotLiveBackground();
      }

      return false;
    }
  },

  setLiveBackground() {
    const playbackSpeedDropdown = document.getElementsByClassName("mdc-select__native-control");
    this.liveButton.style.backgroundColor = "Crimson";
    this.liveText.style.cssText = "color:Crimson !important; font-weight: bold;";
    playbackSpeedDropdown[0].setAttribute("disabled", "");
    playbackSpeedDropdown[0].classList.add("disabledDropdown");
  },

  setNotLiveBackground() {
    const playbackSpeedDropdown = document.getElementsByClassName("mdc-select__native-control");
    this.liveButton.style.backgroundColor = "Black";
    this.liveText.style.cssText = "color:Black !important; font-weight: normal;";
    playbackSpeedDropdown[0].removeAttribute("disabled");
    playbackSpeedDropdown[0].classList.remove("disabledDropdown");
  },

  goToLive() {
    if (!this.islive) {
      const liveTime = new Date(getLiveTime());
      this.islive = true;
      this.setLiveBackground();
      storage.dateKeeper$.period = 1000;
      storage.dateKeeper$.setNewTime(liveTime);
    }
  }
};