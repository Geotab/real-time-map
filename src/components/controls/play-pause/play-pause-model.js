import storage from "../../../dataStore/index";

export const pausePlayModel = {
  playBackBox: undefined,
  playing: true,
  pausePlayIcon: undefined,

  initPausePlay() {
    this.pausePlayIcon = document.getElementById("RTMControlButton");
    this.pausePlayIcon.classList.add("RTM-pauseIcon");
    storage.dateKeeper$.subscribe(this.updatePausePlay.bind(this));
  },

  updatePausePlay() {
    if (this.playing === storage.dateKeeper$.paused) {
      this.playing = !this.playing;
      if (this.playing) {
        this.setPlayBackground();
      } else {
        this.setPausedBackground();
      }
    }
  },

  togglePausePlay() {
    if (this.playing) {
      this.setToPause();
    } else {
      this.setToPlay();
    }
  },

  setToPlay() {
    this.playing = true;
    this.setPlayBackground();
    storage.dateKeeper$.resume();
  },

  setToPause() {
    this.playing = false;
    this.setPausedBackground();
    storage.dateKeeper$.pause();
  },

  setPlayBackground() {
    this.pausePlayIcon.classList.remove("RTM-playIcon");
    this.pausePlayIcon.classList.add("RTM-pauseIcon");
  },

  setPausedBackground() {
    this.pausePlayIcon.classList.remove("RTM-pauseIcon");
    this.pausePlayIcon.classList.add("RTM-playIcon");
  }
};
