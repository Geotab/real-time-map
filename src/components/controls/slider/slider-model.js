import noUiSlider from "propellerkit-range-slider/node_modules/nouislider/distribute/nouislider";
import storage from "../../../dataStore";

import {
  checkIfLive,
  getLiveTime,
  resetTransitionAnimation,
} from "../../../utils/helper";

import {
  getPixelsPerMinute,
  minutesInDay,
  getMinuteOfDay,
  calulateCurrentTime,
  getTimeInlocalFormat,
  minuteOfDayToHour,
  calculateLeftOffset
} from "./slider-helper";

export const sliderModel = {

  userCurrentlySlidng: false,
  currentHandle: 2,

  range: {
    "min": 0,
    "max": minutesInDay
  },

  get RTMSlider() {
    return document.getElementById("RTM-TimeSlider");
  },

  initSlider() {

    const endHandleValue = getMinuteOfDay(storage.currentTime);

    const toolTipFormatter = {
      //Display time in us format.
      from: getTimeInlocalFormat,
      to: getTimeInlocalFormat
    };

    noUiSlider.create(this.RTMSlider, {
      step: 1, //Each step is one minute. There are 1440 minutes a day.
      start: [getMinuteOfDay(storage.timeRangeStart), endHandleValue],
      range: this.range,
      pips: {
        mode: "values",
        values: [...Array(8).keys()].slice(1).map(val => Math.round(val * 180)),
        density: 180,
        format: {
          // 'to' the formatted value. Receives a number.
          to: minuteOfDayToHour,
        },
        stepped: true
      },
      behaviour: "drag-tap",
      connect: true,
      tooltips: [toolTipFormatter, toolTipFormatter]
    });

    this.RTMSlider.noUiSlider.on("start.one", this.startSliding.bind(this));
    this.RTMSlider.noUiSlider.on("slide.one", this.userSliding.bind(this));
    this.RTMSlider.noUiSlider.on("set.one", this.sliderValueSet.bind(this));

    storage.dateKeeper$.subscribe(this.updateSlider.bind(this));
    this.createSliderTitles();
  },

  createSliderTitles() {
    const lower = document.getElementsByClassName("noUi-handle-lower");
    lower[0].title = "Start Time";
    const upper = document.getElementsByClassName("noUi-handle-upper");
    upper[0].title = "Current Time";
  },

  updateSlider(currentSecond) {

    if (!this.userCurrentlySlidng) {
      const endHandleValue = getMinuteOfDay(currentSecond);
      const start = [getMinuteOfDay(storage.timeRangeStart), endHandleValue];

      const options = {
        range: this.range,
        padding: [0, endHandleValue + 1],
        start
      };

      this.RTMSlider.noUiSlider.updateOptions(options, false);

      this.handleOverFlowFix(start, 1);
      this.toolTipOverFlowFix(start, 1);
    };
  },

  startSliding(values, handle) {
    this.toolTipOverFlowFix(values, handle);
    this.currentHandle = handle;
    this.userCurrentlySlidng = true;
  },

  userSliding(values, handle) {
    if (this.currentHandle === 2) {
      this.handleOverFlowFix(values, handle);
    }
    this.toolTipOverFlowFix(values, handle);
    this.userCurrentlySlidng = true;
  },

  sliderValueSet(values, handle) {
    // handle 0 = first handle, 1 = second handle
    this.userCurrentlySlidng = false;
    this.handleOverFlowFix(values, handle);

    storage.timeRangeStart = calulateCurrentTime(values[0]);
    if (checkIfLive(storage.timeRangeStart) > 600) {
      storage.timeRangeStart = getLiveTime();
    }

    let newTime = calulateCurrentTime(values[1]);
    if (this.currentHandle === 0) {
      storage.dateKeeper$.update();

    } else {
      if (checkIfLive(newTime) > 600) {
        newTime = getLiveTime();
      }
      storage.dateKeeper$.setNewTime(newTime);
    }

    this.currentHandle = 2;
  },

  toolTipOverFlowFix(values, handle) {

    const [lowerValue, upperValue] = values;

    const tooltips = document.getElementsByClassName("noUi-tooltip");
    const [lowerToolTip, upperToolTip] = tooltips;

    const slider = document.getElementById("RTM-TimeSlider");
    const pixelsPerMinute = getPixelsPerMinute(slider.offsetWidth);

    const offsetPixelWidth = 36;

    if (handle === 0) {
      const leftOffset = calculateLeftOffset(offsetPixelWidth, lowerValue, pixelsPerMinute);
      lowerToolTip.style.setProperty("left", `${leftOffset}px`, "important");
    } else {
      const leftOffset = calculateLeftOffset(offsetPixelWidth, upperValue, pixelsPerMinute);
      upperToolTip.style.setProperty("left", `${leftOffset}px`, "important");
    }

  },

  handleOverFlowFix(values) {

    const [lowerValue] = values;

    const slider = document.getElementById("RTM-TimeSlider");
    const pixelsPerMinute = getPixelsPerMinute(slider.offsetWidth);

    const handles = document.getElementsByClassName("noUi-handle");
    const [lowerHandle, upperHandle] = handles;

    const leftOffset = calculateLeftOffset(9, lowerValue, pixelsPerMinute);
    lowerHandle.style.setProperty("left", `${leftOffset - 9}px`, "important");

    if (values[1] === values[0]) {
      upperHandle.style.setProperty("left", `${leftOffset}px`, "important");
    } else {
      upperHandle.style.setProperty("left", "0px", "important");
    }
  },

};