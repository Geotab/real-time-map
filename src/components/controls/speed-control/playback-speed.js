import storage from "../../../dataStore";
import { MDCSelect } from "@material/select";
import { checkIfLive } from "../../../utils/helper";

export const playBackSpeedModel = {
  select: undefined,
  speedUpdated: false,

  initPlayBackSpeed() {

    playBackSpeedModel.select = new MDCSelect(document.querySelector(".mdc-select"));
    playBackSpeedModel.select.listen("MDCSelect:change", this.newSpeedSelected.bind(this));

  },

  newSpeedSelected() {

    if (this.speedUpdated) {
      this.speedUpdated = false;
      return;
    }

    this.speedUpdated = true;

    const timeAhead = checkIfLive(storage.currentTime);
    if (timeAhead) {
      playBackSpeedModel.select.value = 1;
    } else {

      const { value } = playBackSpeedModel.select;
      storage.dateKeeper$.setPeriod(1000 / value);
    }

  },

  updateSpeed(period) {
    this.speedUpdated = true;
    const speed = 1000 / period;
    playBackSpeedModel.select.value = Math.round(speed);
  }
};