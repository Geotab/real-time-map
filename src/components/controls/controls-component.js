import React, { Component } from "react";
import { LiveButtonComponent } from "../controls/live-button-model/liveButton-component";
import { PlayPauseButtonComponent } from "./play-pause/play-pause-component";
import { SpeedControlComponent } from "./speed-control/speed-control-component";
import { sliderModel } from "./slider/slider-model";
import { dateTimeModel } from "../controls/date-time-model";
import { liveButtonModel } from "../controls/live-button-model";
import { pausePlayModel } from "./play-pause/play-pause-model";
import { playBackSpeedModel } from "./speed-control/playback-speed";

export class ControlsView extends React.Component {
  componentDidMount() {
    playBackSpeedModel.initPlayBackSpeed();
    pausePlayModel.initPausePlay();
    sliderModel.initSlider();
    dateTimeModel.initDateTimeInput();
    liveButtonModel.initLiveButton();
  }

  render() {
    return (
      //Control bar at the bottom of the map
      <React.Fragment>
        <div id="RTM-ControlBarContainer">
          <div id="Slider-Container">
            <div id="RTM-TimeSlider" className="pmd-range-slider"></div>
          </div>
          <div id="RTM-ControlsContainer">
            <div id="timeControls">
              <PlayPauseButtonComponent />
              <SpeedControlComponent />
            </div>

            <div className="inputControls ">
              <label> Date: </label>
              <input
                className="timeInputBox mdc-button"
                type="date"
                id="dateInputBox"
                step="1"
              ></input>
            </div>

            <div className="inputControls ">
              <label> Start Time: </label>
              <input
                className="timeInputBox mdc-button"
                type="time"
                id="timeRangeStart"
                step="1"
              ></input>
            </div>

            <div className="inputControls ">
              <label> Current Time: </label>
              <input
                className="timeInputBox mdc-button"
                type="time"
                id="currentTimeInput"
                step="1"
              ></input>
            </div>

            {/* //Jump to live time. */}
            <LiveButtonComponent />
          </div>
        </div>
      </React.Fragment>
    );
  }
}
