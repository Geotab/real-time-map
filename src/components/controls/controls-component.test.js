import React from "react";
import { ControlsView } from "./controls-component";
import { render, mount, shallow } from "enzyme";
import { spyOnFunction } from "../../../tests/utils/test-helpers";
import { sliderModel } from "./slider/slider-model";
import { dateTimeModel } from "../controls/date-time-model";
import { liveButtonModel } from "../controls/live-button-model";
import { pausePlayModel } from "./play-pause/play-pause-model";
import { playBackSpeedModel } from "./speed-control/playback-speed";

test("should successfully render the component and match screenshot", () => {
  spyOnFunction(liveButtonModel, "initLiveButton");
  spyOnFunction(pausePlayModel, "initPausePlay");
  spyOnFunction(sliderModel, "initSlider");
  spyOnFunction(dateTimeModel, "initDateTimeInput");
  spyOnFunction(playBackSpeedModel, "initPlayBackSpeed");
  const component = shallow(<ControlsView />);
  expect(component).toMatchSnapshot();
});
