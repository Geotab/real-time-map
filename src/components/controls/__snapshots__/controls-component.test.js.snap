// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should successfully render the component and match screenshot 1`] = `
<Fragment>
  <div
    id="RTM-ControlBarContainer"
  >
    <div
      id="Slider-Container"
    >
      <div
        className="pmd-range-slider"
        id="RTM-TimeSlider"
      />
    </div>
    <div
      id="RTM-ControlsContainer"
    >
      <div
        id="timeControls"
      >
        <PlayPauseButtonComponent />
        <SpeedControlComponent />
      </div>
      <div
        className="inputControls "
      >
        <label>
           Date: 
        </label>
        <input
          className="timeInputBox mdc-button"
          id="dateInputBox"
          step="1"
          type="date"
        />
      </div>
      <div
        className="inputControls "
      >
        <label>
           Start Time: 
        </label>
        <input
          className="timeInputBox mdc-button"
          id="timeRangeStart"
          step="1"
          type="time"
        />
      </div>
      <div
        className="inputControls "
      >
        <label>
           Current Time: 
        </label>
        <input
          className="timeInputBox mdc-button"
          id="currentTimeInput"
          step="1"
          type="time"
        />
      </div>
      <LiveButtonComponent />
    </div>
  </div>
</Fragment>
`;
